### ISSUE TITLE
Dead code elimination

### ISSUE THREAD

**Issue by ZimM-LostPolygon** [sq/JSIL#332]
Lately, I was working on dead code elimination in JSIL to decrease the enormouse size of core libraries. Right now, it allows to strip out unused classes, methods, and fields. Of course, it breakes the reflection for everything that was stripped out, but it is possible to avoid that by whitelisting class/member in cofiguration. 
For a simple console "hello world" app Mono corlib is reduced from 5 MB to 300 kB, and MS corlib reduced from 9.7 MB to 1.3 MB (that's still way too large for a hello-world, have to investigate that further).

It works fairly well for a prototype, but I was thinking - is this really a job JSIL must do itself? Maybe it would make more sense to make stripper in a separate project and just use it as external tool from JSIL. What do you think?

**Comment by Rohansi** [comment]
The DCE can eliminate a lot more if it handled static constructors properly. Right now it basically treats all of them as an entry point, even if the type isn't used.

I implemented most if not all of the cases where static constructors are called and the .NET mscorlib was reduced from 3 MB (current DCE) to 12.6 KB for a console application that prints "hello".

**Comment by LostPolygon** [comment]
Does it? As far as I remember, static constructors are only analyzed when they the type is encountered the first time, so definitely not treating every static constructor as an entry point.

**Comment by Rohansi** [comment]
Sorry, I wasn't handling attributes properly. Some static constructors on attributes bring in a ton of stuff. 

It might be worth having flags to disable stuff like this. If you aren't going to be reading attributes at runtime there's no point having that extra 3 MB.

**Comment by kg** [comment]
Ah, the problem is that a method's attributes pull stuff in?

Maybe you could special-case it to only do the attribute analysis if the function uses the reflection APIs?

**Comment by iskiselev** [comment]
Probably we should check, if JSIgnore could be used with attributes and then provide some config that will strip all BCL attributes?

**Comment by Rohansi** [comment]
Stripping attributes could work but it doesn't fit into the analyzer API.

More specific proxies would help without causing problems, though. I just added a `JSIgnore` static constructor to `ConsoleProxy`  and it's back down to 13 KB. Looks like I was wrong about the attributes.

**Comment by iskiselev** [comment]
If JSIgnore doesn't work correct with attributes in DCE it sounds strange, as I definitely had checked it. I'll look on it.

**Comment by iskiselev** [comment]
I've tested. Looks like marking Attribute with JSIgnore strip it from output. Here is test case:
```
using System;
using System.IO;
using System.Linq;
using System.Text.RegularExpressions;
using System.Threading.Tasks;

using JSIL.Meta;

[TestAttribute]
public static class Program
{

    public static void Main(string[] args)
    {

    }

}

[JSIgnore]
public class TestAttribute : Attribute
{
    public TestClass TestValue;

    public TestAttribute()
    {
        TestValue = new TestClass();
    }
}

public class TestClass
{
    public string TestValue;
    public TestClass()
    {
        TestValue = "tt";
    }

}
```
It even hides another JSIL problem - attributes, marked with `JSIgnore` are stripped from output, but  points where they are applied to another members - are not stripped.

Also, while testing this case, I've found another bug. In next example, TestAttribute is stripped with DCE:
```
[TestAttribute]
public static class Program
{

    public static void Main(string[] args)
    {

    }

}

public class TestAttribute : Attribute
{
}
```

Probably it is even good behavior - if only usage of attribute is marking with it, and never type token is loaded - we may strip it or user should add it to white list. On other hand, if we add constructor to the same attribute that will set some filed - it is not stripped. We should check why behavior is different and discuss which is correct.

**Comment by Rohansi** [comment]
> It even hides another JSIL problem - attributes, marked with JSIgnore are stripped from output, but points where they are applied to another members - are not stripped.

This is what I was talking about. The analyzer API only lets you say types/methods/fields are not used, it doesn't let you remove attributes from definitions. JSIL could call MemberCanBeSkipped on them to check if the attribute type was removed, which should fix that.

> In next example, TestAttribute is stripped with DCE:

This is because entry points start directly at the method and since Main doesn't refer to Program, AddType is never called for it, so no attributes are seen.

**Comment by Rohansi** [comment]
What was the reasoning behind [always including static constructors](https://github.com/sq/JSIL/blob/27f0952c9af3c6b5eae4b83169846c28d80e87ef/JSIL/AssemblyTranslator.cs#L1228-L1229)? It doesn't really cause problems but shouldn't that be the analyzer's decision?

**Comment by LostPolygon** [comment]
Hm, that indeed looks strange as we have that in the analyzer as well:
https://github.com/sq/JSIL/blob/master/Compiler/Analyzers/DeadCodeAnalyzer/DeadCodeInfoProvider.cs#L195

**Comment by iskiselev** [comment]
As far as I remember, JSIL always generate fake .cctor instead of original one to initialize static fields.
That's why, DCE check that accepts MemberReference will always return false.

But, we definitely know, that if type is not excluded fully, we must generate its static constructor. It won't affect possibility to exclude full type if there is no reference to it in all other on-excluded parts of program.

Check inside DCE force DCE to include every member, that is referenced inside static constructor of type if there is any reference to that type (as otherwise there will be no reference to static constructor - it is called be environment, not by user).

**Comment by Rohansi** [comment]
But if nothing references the static fields they don't need to be initialized. It's probably there for correctness because it should be called when accessing members on type and it could have side effects. But this means you can't make an analyzer drop some static constructors because they will still be generated and could access members that were removed.

**Comment by iskiselev** [comment]
The problem here is that current implementation of DCE doesn't analyze side effects. If anybody will implement it, we may improve DCE by stripping on which we only set values - not only in static constructor, but everywhere in our application. So, I think on current implementation, such work with static constructors is correct. Really, there is no difference with case when instance constructor initialize some members, but we've stripped any method that read from this members.

**Comment by iskiselev** [comment]
As I know, JSIL itself has some logic for detection side-effect free methods. If it can be used (or updated so that it can be used) in such way, that it will provide information if any method free of side effects (really, we should understand which we mean under side-effect here - is instance method on class that modify only that class side effect free?) - we could try to improve our very simple DCE logic.

**Comment by Rohansi** [comment]
There's nothing really wrong with how the current DCE handles static constructors, it may be too generous but that's fine. The problem is that the choice can't be made by a DCE because it never asks.

**Comment by iskiselev** [comment]
@Rohansi, we could simply change `AssemblyTranslator.ShouldSkipMember` (but DCE in that case should process .cctor members in a little bit different way - as it will be non-original Member, or we could tweak some other part of JSIL so that it will use correct MemberInfo to ask `ShouldSkipMember`), but I still don't see how can you improve DCE to strip static constructor more aggressive.

**Comment by Rohansi** [comment]
Does the DCE depend on that behavior right now? It looks like it works properly without it. I really don't see the point in JSIL forcing static constructors to be generated, that should be the DCE's decision.

**Comment by iskiselev** [comment]
So, here is a sample of current DCE behavior. Let's look on next sample:
```
using System;
public static class Program
{

    public static void Main(string[] args)
    {
        Console.WriteLine("main");
        //new SampleClass();
    }

}

public class SampleClass
{
    static SampleClass()
    {
        Console.WriteLine("SampleClass cctor");
        new SampleClass2();
    }
}


public class SampleClass2
{
    public SampleClass2()
    {
        Console.WriteLine("SampleClass2 cctor");
    }
}
```

In above example, both `SampleClass` and `SampleClass2` will be totally stripped. You will not see any declarations of this classes - nor their static/instance constructor.

If you will uncomment second line in `Main`, both `SampleClass` and `SampleClass2` will be preserved in output - as we use `SampleClass`, we must translate its static constructor. As static constructor has reference to `SampleClass2` (and its instance constructor), we need preserve `SampleClass2` instance constructor in output.

**Comment by Rohansi** [comment]
Yes, I know how it works right now. But why *shouldn't* we be allowed to eliminate static constructors? I understand that it's not an easy thing to do and I'm not saying the DCE needs to do it, but for some reason JSIL doesn't allow analyzers to do it.

**Comment by iskiselev** [comment]
We may change `AssemblyTranslator.ShouldSkipMember` as soon, as anybody will implement some logic in DCE that will have any benefits of checking for static constructor eliminating. We still need modify `AssemblyTranslator` (see `TranslateTypeStaticConstructor`), that inject some special logic inside class .cctor.
As this is not so easy work and nobody yet suggested any DCE implementation, that achieve any benefits from it - nobody yet done it and I've only inserted simplest hack-implementation for it.
